apps/email/js/html_cache_restore.js

```diff
@@ -102,103 +250,236 @@ window.htmlCacheRestorePendingMessage = [];
     };
   }
 
-  /*
-   * Automatically restore the HTML as soon as module is executed.
-   * ASSUMES card node is available (DOMContentLoaded or execution of
-   * module after DOM node is in doc)
+  /**
+   * Called by app code. Only expects one listener to be registered for each
+   * handler type. This function also assumes that a  `require` loader is
+   * available to fetch the 'evt' module. This would not be needed if
+   * evt.emit('notification') was  not triggered by the email code.
+   * @param  {Object} listener Object whose keys are the handler type names and
+   * values are functions that handle that type.
    */
-  var cardsNode = document.getElementById(selfNode.dataset.targetid);
+  window.globalOnAppMessage = function(listener) {
 
-  function startApp() {
-    var scriptNode = document.createElement('script');
+    Object.keys(listener).forEach(function(key) {
+      var fn = handlers[key] = listener[key];
+      var queue = handlerQueues[key];
+      if (queue.length) {
+        handlerQueues[key] = [];
+        queue.forEach(function(argsArray) {
+          fn.apply(undefined, argsArray);
+        });
+      }
+    });
 
-    if (loader) {
-      scriptNode.setAttribute('data-main', loadSrc);
-      scriptNode.src = loader;
+    // Only need to do this wiring once, but globalOnAppMessage could be called
+    // multiple times.
+    if (!evt) {
+      require(['evt'], function(ev) {
+        evt = ev;
+        evt.on('notification', onNotification);
+      });
+    }
+
+    return startupData;
+  };
+
+  // Attach the hasAccount so that other code can use it and always get the
+  // freshest cached value.
+  window.globalOnAppMessage.hasAccount = hasAccount;
+
+  function dispatch(type, args) {
+    if (handlers[type]) {
+      return handlers[type].apply(undefined, args);
     } else {
-      scriptNode.src = loadSrc;
+      handlerQueues[type].push(args);
     }
 
-    document.head.appendChild(scriptNode);
+    // On the very first dispatch when app open is triggered by a dispatchable
+    // event, need to finish bootup now that full startup state is known.
+    finishStartup();
   }
 
-  // TODO: mozHasPendingMessage can only be called once?
-  // Need to set up variable to delay normal code logic later
-  if (navigator.mozHasPendingMessage) {
-    if (navigator.mozHasPendingMessage('activity')) {
-      window.htmlCacheRestorePendingMessage.push('activity');
+  /**
+   * Perform requested activity.
+   *
+   * @param {MozActivityRequestHandler} req activity invocation.
+   */
+  function onActivityRequest(req) {
+    console.log('mozSetMessageHandler: received an activity');
+    if (messageTooFast('activity') || ignoreMessage('activity')) {
```

Comment:
ignoreMessage should come befor messageTooFast, but its lack of caring about its type/pendingMessage is a little concerning.  Specifically, based on the code as written, if we have both 'notification' and 'activity' pending, 'activity' will win.  But onNotification will fire first and its invocation of messageTooFast() will return false, then ignoreMessage will kill it.  Then onActivityRequest will fire and messageTooFast will return true for it and then things maybe get broken for a while until another activity eventually gets past the messageTooFast gauntlet so ignoreMessage() can have its pendingMessage-clearing side-effect.